spring-cloud-stream/src/test/java/org/springframework/cloud/stream/binder/integration/SpringIntegrationChannelBinder.java

```diff
@@ -0,0 +1,274 @@
+			this.listenerContainer = listenerContainer;
+		}
+
+		@SuppressWarnings("unused")
+		// Temporarily unused until DLQ strategy for this binder becomes a requirement
+		public void setRecoveryCallback(RecoveryCallback<? extends Object> recoveryCallback) {
+			this.recoveryCallback = recoveryCallback;
+		}
+
+		public void setRetryTemplate(RetryTemplate retryTemplate) {
+			this.retryTemplate = retryTemplate;
+		}
+
+		@Override
+		protected void onInit() {
+			if (this.retryTemplate != null) {
+				Assert.state(getErrorChannel() == null, "Cannot have an 'errorChannel' property when a 'RetryTemplate' is "
+						+ "provided; use an 'ErrorMessageSendingRecoverer' in the 'recoveryCallback' property to "
+						+ "send an error message when retries are exhausted");
+			}
+			Listener messageListener = new Listener();
+			if (this.retryTemplate != null) {
+				this.retryTemplate.registerListener(messageListener);
+			}
+			this.listenerContainer.setMessageListener(messageListener);
+		}
+
+		protected class Listener implements RetryListener, Consumer<Message<?>> {
+
+			@Override
+			@SuppressWarnings("unchecked")
+			public void accept(Message<?> message) {
+				try {
+					if (IntegrationBinderInboundChannelAdapter.this.retryTemplate == null) {
+						try {
+							processMessage(message);
+						}
+						finally {
+							attributesHolder.remove();
+						}
+					}
+					else {
+						IntegrationBinderInboundChannelAdapter.this.retryTemplate.execute(context -> {
+									processMessage(message);
+									return null;
+								},
+								(RecoveryCallback<Object>) IntegrationBinderInboundChannelAdapter.this.recoveryCallback);
+					}
+				}
+				catch (RuntimeException e) {
+					if (getErrorChannel() != null) {
+						getMessagingTemplate().send(getErrorChannel(), buildErrorMessage(null,
+								new IllegalStateException("Message conversion failed: " + message, e)));
+					}
+					else {
+						throw e;
+					}
+				}
+			}
+
+			private void processMessage(Message<?> message) {
+				sendMessage(message);
+			}
+
+			@Override
+			public <T, E extends Throwable> boolean open(RetryContext context, RetryCallback<T, E> callback) {
+				if (IntegrationBinderInboundChannelAdapter.this.recoveryCallback != null) {
+					attributesHolder.set(context);
+				}
+				return true;
+			}
+
+			@Override
+			public <T, E extends Throwable> void close(RetryContext context, RetryCallback<T, E> callback,
+					Throwable throwable) {
+				attributesHolder.remove();
+			}
+
+			@Override
+			public <T, E extends Throwable> void onError(RetryContext context, RetryCallback<T, E> callback,
+					Throwable throwable) {
+				// Empty
+			}
+		}
+	}
+
+	/**
+	 * Implementation of ErrorMessageStrategy used by this binder.
+	 * Modeled after the one in Rabbit binder although doesn't do much, so a placeholder for now .
+	 */
+	private static class IntegrationHeaderErrorMessageStrategy implements ErrorMessageStrategy {
```

Comment:
I somehow think the `DefaultErrorMessageStrategy` is enough.
Although I guess you might have something in mind for the future.
Since it is semantically is the same, I won't mind to keep it here just for consistency with Rabbit Binder for example.

Reply:
Ohh, no. That is exactly the feedback I needed as I was more concerned about modeling it after existing Rabbit binder, then the individual components like this one.